Using OIDC authentication for Github action deployment

kdesao-devops opened the pull request:
# Context
We want to use the OIDC module to authenticate the Github action without using terraform cloud.

## Changes
- [x] add Github action permissions
- [x] Add Github environment 
- [x] Introduce `secrets.TERRAFORM_DEPLOY_ROLE_ARN` for each env
- [x] Removing `secrets.TFC_TEAM_TOKEN`
- [x] Update README.md

## Test
- [x] [Deploy](https://github.com/kdesao-devops/startup-sample-project-aws-virtual-machines/actions/runs/3567866382) on Prod
- [x] [Destroy](https://github.com/kdesao-devops/startup-sample-project-aws-virtual-machines/actions/runs/3567876959) on Prod

wrnu commented on .github/workflows/deploy_prod.yml:
```diff
@@ -9,59 +9,50 @@ on:
-            
-            
+            bucket = "${{ secrets.S3_BACKEND_NAME }}"
+            key = "${{ secrets.LICENCEPLATE }}/${{ env.environment }}/serverless-app.tfstate"
```
this is the VM demo app :)

kdesao-devops replied:
This was a poor paste ^^

wrnu commented on .github/workflows/deploy_dev.yml:
```diff
@@ -9,59 +9,50 @@ on:
-            
-            
+            bucket = "${{ secrets.S3_BACKEND_NAME }}"
+            key = "${{ secrets.LICENCEPLATE }}/${{ env.environment }}/serverless-app.tfstate"
```
this is the VM demo app :)

kdesao-devops replied:
This was a poor paste ^^

wrnu commented on README.md:
```diff
@@ -1,12 +1,44 @@
   - `LICENCEPLATE` is the 6 character licence plate associated with your project set e.g. `abc123`
-  - `TFC_TEAM_TOKEN` is the token used to access the terraform cloud runner.
+  - `S3_BACKEND_NAME` is the name of the S3 Bucket name used to store the Terraform state.
+
```
We need docs for  `secrets.TERRAFORM_DEPLOY_ROLE_ARN`

kdesao-devops replied:
It's just on top, isnn't it enough ?

wrnu replied:
Oops sorry, I didn't see it.

wrnu reviewed (changes requested):
Once this is merged we'll need to create environments and set all of the required secrets

kdesao-devops commented on .github/workflows/deploy_test.yml:
```diff
@@ -37,31 +38,23 @@ jobs:
-            
-            
+            bucket = "${{ secrets.S3_BACKEND_NAME }}"
+            key = "${{ secrets.LICENCEPLATE }}/${{ env.environment }}/serverless-app.tfstate"
```
This was a poor paste ^^

kdesao-devops commented:
I have struggled a little bit but I have fixed the PR workflow for Prod.

So I added a job to check the name of the branch and I set an output if main => github_env = `prod` ; else github_env = `base branch`

```
jobs:
  env_selector:
    runs-on: ubuntu-latest       
    steps:
      - name: env selector
        id: identify
        run: if [ ${GITHUB_HEAD_REF} = "main" ];then export github_env='prod'; else export github_env=${GITHUB_HEAD_REF}; fi
    outputs:
      github_env: ${{ steps.identify.outputs.github_env }}

  build:
    needs: env_selector
    environment: ${{ needs.env_selector.outputs.github_env }}
```
